File: Matlab/envelope.m
function [envelope_data] = envelope(train_label, train_data, test_data, num_std)
%==========================================================================
% Inputs:
% train_label           [m x 1] : label for training data
% train_data			[m x p] : training data for sequences
% test_data				[n x p] : testing data for sequences
% num_std				[k] 	: number of standard deviations  
%--------------------------------------------------------------------------
% Outputs:
% envelope_data			[n x ?] : represent testing data with envelope
%==========================================================================
    % Get unique label
    unique_label = unique(train_label);
	
	% Get dimensions
    s_train = size(train_data);
    s_test = size(test_data);
    s_label = size(unique_label);
	
    % Initial mean and std array
    mean_train = zeros(s_label(1), s_train(2));
    std_train = zeros(s_label(1), s_train(2));

	% Compute mean and std separately
    for count=1:s_label(1)
        label = unique_label(count);
        mean_train(count, :) = mean(train_data(train_label==label, :));
        std_train(count, :)  = std(train_data(train_label==label, :));
    end

    size(mean_train)
	% Initial envelope data
    datalen = 3;
    envelope_data = zeros(s_test(1), datalen*s_label(1));
            
            num_one = sum(test_data(i, :)>(mean_train(count, :)+num_std*std_train(count, :)));      
            num_mone = sum(test_data(i, :)<(mean_train(count, :)-num_std*std_train(count, :)));
            
            envelope_data(i, (count-1)*datalen+1) = s_test(2)-num_one-num_mone;                                %(0)
            envelope_data(i, (count-1)*datalen+2) = num_one;                                                                 %(1)
            envelope_data(i, (count-1)*datalen+3) = num_mone;                                                             %(-1)
        end
    end
end

File: Python/Envelope.py
import numpy as np

def envelope(train_label, train_data, test_data, num_std):
    unique_label = set(train_label)
    s_train = [len(train_data), np.max([np.max(map(lambda x: len(x), train_data)),len(test_data[0])])]
    s_test  = [len(test_data),len(test_data[0])]
    s_label = len(unique_label)

    mean_train = np.zeros([s_label, s_train[1]])
    std_train = np.zeros([s_label, s_train[1]])

    for i in range(s_label):
        now_label_data = []
        label = unique_label.pop()
        # Find the same label's data
        for t in range(len(train_data)):
                if train_label[t] == label:
                        now_label_data.append(train_data[t])
        # Compute the mean
        mean_train[i][:len(now_label_data[0])] = np.mean(now_label_data, axis=0)

        # Compute the std
        std_train[i][:len(now_label_data[0])]  = np.std(now_label_data, axis=0)

    envelope_data = np.zeros([s_test[0], 3 * s_label])

    for count in range(0, s_label):
        num_one   = np.sum(test_data > mean_train[count][:len(test_data[0])] + num_std * std_train[count][:len(test_data[0])], 1)
        num_minus = np.sum(test_data < mean_train[count][:len(test_data[0])] - num_std * std_train[count][:len(test_data[0])], 1)
        envelope_data[:, (count)*3] = s_test[1] - num_one - num_minus
        envelope_data[:, (count)*3 + 1] = num_one
        envelope_data[:, (count)*3 + 2] = num_minus

    return envelope_data

if __name__ == '__main__':
    train_label = [1,1,2,2]
    train_data = [[1,0,1], [1,1,1], [0,0,1,1], [1,0,1,1]]
    test_data = [[1,1,1], [1,2,2]]
    num_std = 1
    print envelope(train_label, train_data, test_data, num_std)


